File: core/src/net/natruid/jungle/systems/ViewManageSystem.java
package net.natruid.jungle.systems;

import com.artemis.BaseSystem;
import com.badlogic.gdx.utils.ObjectMap;
import net.natruid.jungle.core.Sky;
import net.natruid.jungle.utils.Utils;
import net.natruid.jungle.views.AbstractView;

@SuppressWarnings("unchecked")
public class ViewManageSystem extends BaseSystem {
    private ObjectMap<Class<? extends AbstractView>, AbstractView> viewMap = new ObjectMap<>();

    public <T extends AbstractView> T show(Class<T> type) {
        AbstractView view = viewMap.get(type);
        if (view == null) {
            try {
                view = type.newInstance();
                viewMap.put(type, view);
            } catch (InstantiationException | IllegalAccessException e) {
                e.printStackTrace();
            }
        }

        assert view != null;
        Sky.jungle.showView(view);
        return (T) view;
    }

    public <T extends AbstractView> T get(Class<T> type) {
        AbstractView view = viewMap.get(type);
        if (view == null) return null;
        return (T) view;
    }

    public <T extends AbstractView> T hide(Class<T> type) {
        T t = get(type);
        if (t != null) Sky.jungle.hideView(t);
        return t;
    }

    public AbstractView hideLast() {
        return Sky.jungle.hideLastView();
    }

    public void hideAll() {
        for (AbstractView view : viewMap.values()) {
            Sky.jungle.hideView(view);
        }
    }

    @Override
    protected void dispose() {
        for (AbstractView view : viewMap.values()) {
            Utils.safelyDispose(view);
        }
        viewMap.clear();
    }

    @Override
    protected void processSystem() {
    }
}

File: core/src/net/natruid/jungle/systems/ViewManageSystem.kt
package net.natruid.jungle.systems

import com.artemis.BaseSystem
import ktx.assets.disposeSafely
import net.natruid.jungle.core.Jungle
import net.natruid.jungle.views.AbstractView
import kotlin.reflect.KClass

@Suppress("UNCHECKED_CAST")
class ViewManageSystem : BaseSystem() {
    private val viewMap = HashMap<KClass<out AbstractView>, AbstractView>()

    inline fun <reified T : AbstractView> show(): T {
        return show(T::class)
    }

    fun <T : AbstractView> show(type: KClass<T>): T {
        var view = viewMap[type]
        if (view == null) {
            view = AbstractView.createView(type)
            viewMap[type] = view
        }
        Jungle.instance.showView(view)
        return view as T
    }

    inline fun <reified T : AbstractView> get(): T? {
        return get(T::class)
    }

    fun <T : AbstractView> get(type: KClass<out AbstractView>): T? {
        val view = viewMap[type] ?: return null
        return view as T
    }

    inline fun <reified T : AbstractView> hide(): T? {
        return hide(T::class)
    }

    fun <T : AbstractView> hide(type: KClass<out AbstractView>): T? {
        val view = viewMap[type] ?: return null
        Jungle.instance.hideView(view)
        return view as T
    }

    fun hideLast(): AbstractView? {
        return Jungle.instance.hideLastView()
    }

    fun hideAll() {
        viewMap.values.forEach {
            Jungle.instance.hideView(it)
        }
    }

    override fun dispose() {
        viewMap.values.forEach {
            it.disposeSafely()
        }
        viewMap.clear()
    }

    override fun processSystem() {}
}
